src/box/lua/experimental/connpool.lua

```diff
@@ -0,0 +1,191 @@
+            connect_timeout = opts.connect_timeout,
+            wait_connected = false,
+            fetch_schema = opts.fetch_schema,
+        }
+        local ok, res = pcall(netbox.connect, uri, conn_opts)
+        if not ok then
+            local msg = 'Unable to connect to instance %q: %s'
+            error(msg:format(instance_name, res.message), 0)
+        end
+        conn = res
+        connections[instance_name] = conn
+    end
+
+    -- If opts.wait_connected is not false we wait until the connection is
+    -- established or an error occurs (including a timeout error).
+    if opts.wait_connected ~= false and conn:wait_connected() == false then
+        local msg = 'Unable to connect to instance %q: connection timeout'
+        error(msg:format(instance_name), 0)
+    end
+    return conn
+end
+
+local function is_roles_match(expected_roles, present_roles)
+    if expected_roles == nil or next(expected_roles) == nil then
+        return true
+    end
+    if present_roles == nil or next(present_roles) == nil then
+        return false
+    end
+
+    local roles = {}
+    for _, present_role_name in pairs(present_roles) do
+        roles[present_role_name] = true
+    end
+    for _, expected_role_name in pairs(expected_roles) do
+        if roles[expected_role_name] == nil then
+            return false
+        end
+    end
+    return true
+end
+
+local function is_labels_match(expected_labels, present_labels)
+    if expected_labels == nil or next(expected_labels) == nil then
+        return true
+    end
+    if present_labels == nil or next(present_labels) == nil then
+        return false
+    end
+
+    for label, value in pairs(expected_labels) do
+        if present_labels[label] ~= value then
+            return false
+        end
+    end
+    return true
+end
+
+local function is_candidate_match(instance_name, opts)
+    assert(opts ~= nil and type(opts) == 'table')
+    local get_opts = {instance = instance_name}
+    return is_roles_match(opts.roles, config:get('roles', get_opts)) and
+           is_labels_match(opts.labels, config:get('labels', get_opts))
+end
+
+local function filter(opts)
+    checks({
+        labels = '?table',
+        roles = '?table',
+    })
+    local candidates = {}
+    for instance_name in pairs(config:instances()) do
+        if is_candidate_match(instance_name, opts or {}) then
+            table.insert(candidates, instance_name)
+        end
+    end
+    return candidates
+end
+
+local function get_connection(all_candidates, prefer_local)
+    local candidates = table.copy(all_candidates)
+    if prefer_local ~= false then
+        local candidate_idx = nil
+        for n, candidate in ipairs(candidates) do
+            if candidate == box.info.name then
+                candidate_idx = n
+                local conn = connect(box.info.name, {wait_connected = false})
+                if conn:wait_connected() then
+                    return conn
+                end
```

Comment:
Is not it the same as just connect with wait_connected = true?

Reply:
I had the same question, but then found that if we'll do `connect()` without options, we need to pcall it to try the next candidate.

Reply:
Also, `connect()` returns an already existing connection, if it exists.